### DISCUSSION TITLE
Expand all relations, and find all parents?

### DISCUSSION THREAD

**Discussion by ccbiozhaw** [pocketbase/pocketbase#2901] [Q&A] [answered]
Hello, I am trying to build an application that would really benefit from a powerfull search feature.
I figured the easiest thing would be to mirror the sql database in meilisearch. Then, every time a user creates data/updates data/deletes data (i dont expect this to happen too often), i also mirror it in meilisearch. 

I found the documentation for event hooks, there are hooks like ```OnCollectionAfterCreateRequest``` which get me the current record.
However, I would like to expand relations aswell (at least one level deep).
is there a way to get a record:

```
R1 {
...
"name": "pocketbase",
"relation1": "10721874497234723047",
"relation2": ["j'49324^239jd", "42'3498392"],
...
}
```

and expand it to 

```
R1 {
...
"name": "pocketbase",
"relation1": {something: "expanded"},
"relation2": [{something: "expanded"}, {something: "expanded"}],
...
}
```

also since i intend to expand relations, it would make sense to additionally update all parent components. so if R1 changes i need to get all records that reference R1, then expand them aswell to include the new R1. thanks

**Comment by ganigeorgiev** [comment, marked as answer]
Note that the `OnCollectionAfterCreateRequest` hook is triggered only when creating new collections, not record data.

For record data changes you can attach to the [`OnRecord*Request` hooks](https://pocketbase.io/docs/event-hooks/#record-crud-api-hooks) or use the [`OnModel*` hooks](https://pocketbase.io/docs/event-hooks/#db-hooks) if you want to listen to every db model change event (not just from API requests).

Additionally, also keep in mind that usually the relation expansion is dependent on whether the request client has View API rule access to the expended collection record(s) so I'm not sure in your case what is the desired behavior. If you want to ignore the per-client access check and always have expanded relations with the record in your external service, you can use the [`Dao.ExpandRecord`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.8/daos#Dao.ExpandRecord) and [`Dao.ExpandRecords`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.8/daos#Dao.ExpandRecords) helpers, something like:

```go
app.OnRecordAfterCreateRequest("collection1", "collection2").Add(func(e *core.RecordCreateEvent) error {
	// side-note: you can also execute the below in a new
	// goroutine (aka. go func) to avoid blocking the main request
	// ---

	copy := e.Record.CleanCopy() // create a copy to avoid modifying the resolved record response

	app.Dao().ExpandRecord(
		copy,
		[]string{"yourRelField1", "yourRelField2"},
		func(relCollection *models.Collection, relIds []string) ([]*models.Record, error) {
			// return all records
			return app.Dao().FindRecordsByIds(relCollection.Id, relIds, nil)
		},
	)

	// copy now should contain the "yourRelField1" and "yourRelField2" expands
	// 
	// write the copy in your meilisearch db...

	return nil
})
```

**Comment by ccbiozhaw** [comment]
```
schema := copy.Collection().Schema.Fields()

var relations []string
for _, r := range schema {
	if r.Type == "relation" {
		relations = append(relations, r.Name)
	}
}
```

thank you. i can get the fields that could be expanded like this, then pass them to ```app.Dao().ExpandRecord()```

however, is there a way to now only get the relevant data from ```copy``` ? I found the method ```copy.SchemaData()```, however this discards the expanded fields again.

thank you !

**Reply by ganigeorgiev** [reply]
Could you elaborate what do you mean by _"get the relevant data from copy"_, aka. what data do you want to retrieve?

**Reply by ccbiozhaw** [reply]
this is what the copy object looks like: 

![image](https://github.com/pocketbase/pocketbase/assets/80820813/9c04c535-349f-4dc5-850c-62e5ce7af0bf)

what i would prefer is just the data in the record:

this is the output from copy.SchemaData():

```
map[string]interface {}{
    "ddd":    "",
    "field":  "hello",
    "field1": []string{},
    "field2": "ugrx2dlt4327fzq",
}
```

**Reply by ganigeorgiev** [reply]
Do you mean that you want to print also the expand? If that's the case, you can try with [`Record.PublicExport()`](https://pkg.go.dev/github.com/pocketbase/pocketbase/models#Record.PublicExport). If you just want the expands, then you can use [`Record.Expand()`](https://pkg.go.dev/github.com/pocketbase/pocketbase/models#Record.Expand).

**Reply by ccbiozhaw** [reply]
I think ```.PublicExport``` does not propagate to the extended fields. If i ```copy.PublicExport```, i still get Information about the Collection and Schema  in the expanded fields.

**Reply by ganigeorgiev** [reply]
I'm really confused. Each record model has a pointer to its collection. The `expand` field has type `map[string]any`.

The above screenshot is printing the Go struct. Are you trying to generate the JSON representation? If that's the case, then you can call `MarshalJSON()`.

**Reply by ccbiozhaw** [reply]
Hey, thank you for the help. Yea with MarshalJSON() it gives me the representation i am looking for! 

What i am trying to do is take the record -> expand all fields -> create an object i can store in Meilisearch. As such, all of the additional information on the pocketbase Record is not neccessary (like the Schema). I think ```record.SchemaData()``` i get even closer than using ```MarshalJSON()```, however it does not work on the expanded fields, so I have to iterate over them manually ( i dont really need fields like CollectionId, timestamps and stuff either).
